worker/uniter/timer.go

```diff
@@ -4,21 +4,26 @@
 package uniter
 
 import (
+	"math/rand"
 	"time"
-)
 
-const (
-	// interval at which the unit's status should be polled
-	statusPollInterval = 5 * time.Minute
+	"github.com/juju/juju/worker/uniter/remotestate"
 )
 
-// updateStatusSignal returns a time channel that fires after a given interval.
-func updateStatusSignal() <-chan time.Time {
-	// TODO(fwereade): 2016-03-17 lp:1558657
-	return time.After(statusPollInterval)
-}
+// NewUpdateStatusTimer returns a func returning timed signal suitable for update-status hook.
+func NewUpdateStatusTimer() remotestate.UpdateStatusTimerFunc {
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	return func(wait time.Duration) func() <-chan time.Time {
+		// Actual time to wait is randomised to be +/-20%
+		// of the nominal value.
+		lower := 0.8 * float64(wait)
+		window := 0.4 * float64(wait)
+		offset := float64(r.Int63n(int64(window)))
+		wait = time.Duration(lower + offset)
```

Comment:
This seems to randomize the time that any given unit will wait, but not randomize the time that any given wait will trigger. Is that correct?

As in, with this in play, eg mysql/0 will wake up every 3m and trigger, while mysql/1 will wake up every 6m and trigger.

I'd rather see it where on every sleep cycle we recompute a variable wait.
So that mysql/0 will wait 3, then 6, then 4, then 5, etc. That means that all units still average 5min wait times (or whatever was nominally set), but it should still push out when the explicit ticks are happening.

Or maybe that's what you're doing and I'm just confused by the double-nested func returning a func.
It *feels* like we just want to move the recalculation of specific wait time into the second func rather than this first one.